File: simulator/src/edu/berkeley/sixtus/simul/SimulatorEngine.java
package edu.berkeley.sixtus.simul;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Random;

/**
 * 
 * @author xvilajosana
 * 
 */
public class SimulatorEngine {

	public static final int MAX_NET_SIZE = 50;
	public static final int MAX_NUM_NEIGHBORS = 10;
	public static final int MIN_NUM_NEIGHBORS = 2;

	public static final int MAX_TIME_SLOT = 101;
	public static final int MAX_CH_OFFSET = 16;

	public boolean[][] network = new boolean[MAX_NET_SIZE][MAX_NET_SIZE];// matrix matching neighbours
	public Node[] nodes = new Node[MAX_NET_SIZE]; //the nodes
	
	private int num_links_requested;

	public SimulatorEngine() {
      super();
	}

	/**
	 * requires the number of links to be allocated to each neigbour
	 * 
	 * @param links
	 */
	public void simulate(int links) {
		Random ran = new Random();
		this.num_links_requested = links;

		initializeNetwork();

		// create the network randomly.
		createNetworkTopology(ran);
		// compute how many neighbors has each node and update each neighbor with that info
		computeNumNeighbors();

		// for each node in the network request several links to all neighbours
		for (int numlink = 1; numlink < this.num_links_requested; numlink++) {
			System.out.println("************************ requesting " + numlink + " links");
			allocateLinks(ran, numlink);
			printResult(numlink);
			// reset the nodes: for the next iteration
			resetNodes();
			// compute how many neighbors has each node and update each neighbor with that info
			computeNumNeighbors();
		}
	}

	private void allocateLinks(Random ran, int l) {
		for (int k = 0; k < l; k++) {
			for (int i = 0; i < MAX_NET_SIZE; i++) {
				for (int j = 0; j < MAX_NET_SIZE; j++) {
					if (network[i][j]) {
						// these are neighbors
						int slotNumber = ran.nextInt(MAX_TIME_SLOT);// pick random ts and ch.offset
						int channelOffset = ran.nextInt(MAX_CH_OFFSET);
						//TODO, check result of the operation and if false then do not reschedule or reset schedule
						if (nodes[i].isLinkAvailable(slotNumber,channelOffset,Cell.SlotType.TX, j) &&
								nodes[j].isLinkAvailable(slotNumber,channelOffset,Cell.SlotType.TX, j)){
							nodes[i].scheduleLink(slotNumber, channelOffset,Cell.SlotType.TX, j);
							nodes[j].scheduleLink(slotNumber, channelOffset,Cell.SlotType.RX, i);
						}
					}
				}
			}
		}
	}

	private void createNetworkTopology(Random ran) {
		for (int i = 0; i < MAX_NET_SIZE; i++) {
			// select number of neighbors for that node
			int num_nei = ran.nextInt(MAX_NUM_NEIGHBORS);
			if (num_nei < MIN_NUM_NEIGHBORS) {
				num_nei = MIN_NUM_NEIGHBORS;
			}
			// set them in the network structure. pick them randomly too.
			for (int j = 0; j < num_nei; j++) {
				int h = ran.nextInt(MAX_NET_SIZE);
				while (h == i) {
					h = ran.nextInt(MAX_NET_SIZE);// we don't want to say that
													// we are neighbours of
													// ourselves
				}
				network[i][h] = true;// set it as the neighbour
				network[h][i] = true;// set it as the neighbour
			}
		}
	}

	private void initializeNetwork() {
		for (int i = 0; i < MAX_NET_SIZE; i++) {
			for (int j = 0; j < MAX_NET_SIZE; j++) {
				network[i][j] = false; // initialize network
			}
		}
		resetNodes();
	}

	private void resetNodes() {
		for (int i = 0; i < MAX_NET_SIZE; i++) {
			nodes[i] = new Node(i);
		}
	}

	private void printResult(int numlink) {
		String head=null;
		
		if (numlink==1){
			head = "Node,NumLinksRequested,NumNeighbors,TotalLinks,Allocated Links,Collisions,IdenticaAllocation,% Collision,% Used Links";
			System.out.println(head);
		}
			
		for (int i = 0; i < MAX_NET_SIZE; i++) {
			String content = i
					+ ","
					+ numlink
					+ ","
					+ nodes[i].getNumNeighbors()
					+ ","
					+ (MAX_TIME_SLOT * MAX_CH_OFFSET)
					+ ","
					+ nodes[i].getNumAllocated()
					+ ","
					+ nodes[i].getNumCollisions()
					+ ","
					+ nodes[i].getNumIdenticallyAllocated()
					+ ","
					+ ((((double) nodes[i].getNumCollisions()) / ((double) nodes[i].getNumAllocated())) * 100.0)
					+ ","
					+ (((double) nodes[i].getNumAllocated() / (MAX_TIME_SLOT * MAX_CH_OFFSET)) * 100.0);
			
			System.out.println(content);
			
			this.writeToFile("results" + i,head, content);
		}
	}

	public void computeNumNeighbors() {
		for (int i = 0; i < MAX_NET_SIZE; i++) {
			int count = 0;
			for (int j = 0; j < MAX_NET_SIZE; j++) {
				if (network[i][j])
					count++;
			}
			nodes[i].setNumNeighbors(count);
		}
	}

	public void writeToFile(String filename,String header, String content) {
		FileOutputStream fop = null;
		File file;

		try {
			file = new File(filename);

			if (!file.exists()) {
				file.createNewFile();
			}

			fop = new FileOutputStream(file, true);

			if (header!=null) content = header + '\n' + content;
			
			// get the content in bytes
			byte[] contentInBytes = content.getBytes();

			fop.write(contentInBytes);
			fop.write('\n');
			fop.flush();
			fop.close();

		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			try {
				if (fop != null) {
					fop.close();
				}
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}
}

File: python/ScheduleSimulator/simulator/SimulatorEngine.py
'''
Created on 27/06/2013

@author: xvilajosana
'''

import random
import Cell
import Node
import Constants as c

class SimulatorEngine(object):
    '''
    classdocs
    '''

    def __init__(self):
        '''
        Constructor
        '''
        self.network = [[False for x in xrange(c.MAX_NET_SIZE)] for x in xrange(c.MAX_NET_SIZE)] 
    
        
        
    def simulate(self,links):
        
        self.initializeNetwork()
        self.createNetworkTopology()
        self.computeNumNeighbors()
        
        for numLink in range(links):
            print "** requesting Link {0}".format(numLink)
            self.allocateLinks(numLink)
            self.printResult(numLink)
            self.resetNodes()
            self.computeNumNeighbors()
            
    def allocateLinks(self,numLink):
        for k in range(numLink):
            for i in range(c.MAX_NET_SIZE):
                for j in range(c.MAX_NET_SIZE):
                    if self.network[i][j]==True:
                        slotNumber=random.randint(0,c.MAX_TIME_SLOT-1)
                        chOffset=random.randint(0,c.MAX_CH_OFFSET-1)
                        #print "slotNum={0}, chOff={1}".format(slotNumber,chOffset)
                        
                        if ((self.nodes[i].isLinkAvailable(slotNumber,chOffset,c.CELL_TYPE_TX,j) and self.nodes[j].isLinkAvailable(slotNumber,chOffset,c.CELL_TYPE_RX,i))):
                            self.nodes[i].scheduleLink(slotNumber, chOffset,c.CELL_TYPE_TX, j)
                            self.nodes[j].scheduleLink(slotNumber, chOffset,c.CELL_TYPE_RX, i)
                               
                            
                                                        
    def createNetworkTopology(self):
        for i in range(c.MAX_NET_SIZE):
            # select number of neighbors for that node
            num_nei = random.randint(c.MIN_NUM_NEIGHBORS,c.MAX_NUM_NEIGHBORS-1)
            # set them in the network structure. pick them randomly too.
            for j in range(num_nei):
                h = random.randint(0,c.MAX_NET_SIZE-1)
                while (h == i):
                    h = random.randint(0,c.MAX_NET_SIZE-1)# we don't want to say that we are neighbours of ourselves
                self.network[i][h] = True;# set it as the neighbour
                self.network[h][i] = True;# set it as the neighbour

   
   
   
    def initializeNetwork(self):
        for i in range(c.MAX_NET_SIZE):
            for j in range(c.MAX_NET_SIZE):
                self.network[i][j] = False # initialize network
        self.resetNodes();

    def resetNodes(self):
        self.nodes = [Node.Node(count) for count in range(c.MAX_NET_SIZE)]
        


    def printResult(self,numlink):
        head=None

        if numlink==1:
            head = "Node,NumLinksRequested,NumNeighbors,TotalLinks,Allocated Links,Collisions,IdenticaAllocation,% Collision,% Used Links"
            print head

        for i in range(c.MAX_NET_SIZE):
            content = "{0},{1},{2},{3},{4},{5}".format(numlink,self.nodes[i].getNumNeighbors(),c.MAX_TIME_SLOT * c.MAX_CH_OFFSET,self.nodes[i].getNumAllocated(),self.nodes[i].getNumCollisions(),self.nodes[i].getNumIdenticallyAllocated())
            
            print content
            
            
            
    def computeNumNeighbors(self):
        for i in range(c.MAX_NET_SIZE):
            count = 0
            for j in range(c.MAX_NET_SIZE):
                if self.network[i][j]==True:
                    count=count+1
            self.nodes[i].setNumNeighbors(count);
